### PR TITLE
Implement yaml for generate_figure

### LINKED ISSUE #125: Allow exclusion of subject/site using yaml for generate_figure.py
Currently, subjects and sites to exclude from generating figures are defined directly inside [generate_figure.py](https://github.com/spine-generic/spine-generic/blob/master/spinegeneric/cli/generate_figure.py) script. It would be better to pass this information as input yaml config file.

### PR DESCRIPTION
**Pull request by alexfoias**
To test you'll need the https://github.com/spine-generic/data-multi-subject/releases/download/r20200722/sct_run_batch_20200722022726.zip

To test it out:
```
git checkout af/yml_generate_figure

add 'sapienza': 'italy', under flags in generate_figure.py

python spine-generic/spinegeneric/cli/generate_figure.py -config PATH_TO_EXCLUDE.yml -path-results PATH_TO_DATA/sct_run_batch_20200722022726/results
```

Example content of EXCLUDE.yml:
```
csa_t1:
  - sub-oxfordFmrib04
  - sub-mountSinai03
csa_t2:
  - sub-oxfordFmrib04
  - sub-mountSinai03
dti_fa:
  - sub-beijingPrisma03
  - sub-mountSinai03
  - sub-oxfordFmrib04
  - sub-oxfordFmrib01
mtr:
  - sub-beijingPrisma04
  - sub-geneva02
  - sub-oxfordFmrib04
  - stanford
mtsat:
  - sub-geneva02
  - sub-oxfordFmrib04
  - sub-tehranS04
t1:
  - sub-oxfordFmrib04
  - sub-fslAchieva03
  - sub-fslAchieva04
  - sub-fslAchieva05
  - sub-fslAchieva06
```

Fixes https://github.com/spine-generic/spine-generic/issues/125

Other improvements include:
- slight modification to -indiv-subj CLI function (no more 0/1 choice, but store_true)

### CODE REVIEW

**Review by jcohenadad — CHANGES_REQUESTED** [review]
could you also update the documentation on RTD

**Comment by valosekj** [comment]
How the `EXCLUDE.yml` config file will be shared/distributed? Maybe, it could be packaged with the dataset.

**Comment by jcohenadad** [comment]
> Maybe, it could be packaged with each version of the dataset.

yes, that makes sense. But i'm not sure what's the best approach: package it with the dataset or with the spine-generic repos (because this is where the analysis code is located, so it makes more sense to me). In the latter case we would call it e.g. `exclude_data_multi_subject.yml`

**Comment by alexfoias** [comment]
> could you also update the documentation on RTD

Fixed in https://github.com/spine-generic/spine-generic/pull/168/commits/1450830603b0563ea8d45bbae02be4711ba3ec26

File: docs/analysis_pipeline.rst [hunk 1]
```diff
@@ -187,9 +187,10 @@ Generate figures
 ----------------
 
 Generate figures based on the output csv files using ``sg_generate_figures.py`` script. Run this script in ``/results``
-folder (folder containing csv files) or specify this folder using ``-path-results`` flag. Figures will be created in the
-folder `results/`:
+folder (folder containing csv files) or specify this folder using ``-path-results`` flag. The flag ``-config`` is
+mandatory and an yml file containing the subjects/sites that want to be excluded have to be provided. Figures will
```
Comment (jcohenadad) [inline docs/analysis_pipeline.rst:191]:
```suggestion
mandatory and a yml file containing the subjects/sites that want to be excluded have to be provided. Figures will 
```

Comment (jcohenadad) [inline docs/analysis_pipeline.rst:191]: why making it mandatory? what if there is no subject to exclude?

Comment (alexfoias) [reply docs/analysis_pipeline.rst]: I removed it from mandatory

Comment (alexfoias) [reply docs/analysis_pipeline.rst]: Fixed in https://github.com/spine-generic/spine-generic/pull/168/commits/df7ba948a6d791624ebc3fd9abeec4904cc1403a

**Comment by valosekj** [comment]
> > Maybe, it could be packaged with each version of the dataset.
> 
> yes, that makes sense. But i'm not sure what's the best approach: package it with the dataset or with the spine-generic repos (because this is where the analysis code is located, so it makes more sense to me).

Hard to say. If the dataset is always analyzed using spine-generic repo, then yml file can be packaged with spine-generic repo. On the other hand, multi-subject dataset can be theoretically analyzed completely by some another pipeline or SW. Then, it would be better to have information about subjects with low data quality/artifacts in form of yml file directly in dataset repo.

**Comment by jcohenadad** [comment]
>  On the other hand, multi-subject dataset can be theoretically analyzed completely by some another pipeline or SW. Then, it would be better to have information about subjects with low data quality/artifacts in form of yml file directly in dataset repo.

good point, but the yml file is also specific to spine-generic, i.e. the fields (`csa_t1 `, `mtr `,  etc.) refer to specific analysis scripts that are in spine-generic. Well, let's break the tie and put it under the data repository. I don't feel strongly about it. Let's see how it works and we can always revisit.

**Comment by alexfoias** [comment]
@jcohenadad Do you want me to open a new PR to add the yml to data-multi-subject ?

File: spinegeneric/cli/generate_figure.py [hunk 2]
```diff
@@ -467,6 +470,19 @@ def main():
...
+    else:
+        sys.exit("ERROR: Input yml file {} does not exist or path is wrong.".format(args.config))
+
+    # fetch input yml file as dict
+    with open(fname_yml, 'r') as stream:
+        try:
+            dict_exclude_subj = yaml.safe_load(stream)
+        except yaml.YAMLError as exc:
+            print(exc)
+
```
Comment (valosekj) [inline spinegeneric/cli/generate_figure.py:485]: Maybe, we could add condition for initialization of empty `dict_exclude_subj` dict, when no input yml config file is passed (I know, that in most cases some subjects will be excluded and yml file will be passed, but for sure, it is better to deal with situation when no subject will be excluded):

```suggestion
    # create dict with subjects to exclude if input yml config file is passed
    if args.config is not None:
        # check if input yml file exists
        if os.path.isfile(args.config):
            fname_yml = args.config
        else:
            sys.exit("ERROR: Input yml file {} does not exist or path is wrong.".format(args.config))

        # fetch input yml file as dict
        with open(fname_yml, 'r') as stream:
            try:
                dict_exclude_subj = yaml.safe_load(stream)
            except yaml.YAMLError as exc:
                print(exc)
    else:
        # initialize empty dict if no config yml file is passed
        dict_exclude_subj = dict()

```

Comment (jcohenadad) [reply spinegeneric/cli/generate_figure.py:485]: great suggestion @valosekj . i need to resolve the conflicts first, then i'll commit your suggestion

File: spinegeneric/cli/generate_figure.py [hunk 3]
```diff
@@ -38,41 +39,6 @@
 hdlr = logging.StreamHandler(sys.stdout)
 logging.root.addHandler(hdlr)
 
-# List subject to remove, associated with contrast
-SUBJECTS_TO_REMOVE = [
-    # CSA
-    {'subject': 'sub-oxfordFmrib04', 'metric': 'csa_t1'},  # T1w scan is not aligned with other contrasts (subject repositioning)
```
Comment (jcohenadad) [inline spinegeneric/cli/generate_figure.py:44]: we need to keep the comments (ie: reason for removing the subject).
fortunately, yaml supports comments, so these should be ported to the new yaml files.
@valosekj would you mind generating this yaml file and i will upload it along with the database? thanks 🙏

Comment (valosekj) [reply spinegeneric/cli/generate_figure.py:44]: Sure 👍 I have created this file with comments and opened PR - https://github.com/spine-generic/data-multi-subject/pull/11

File: docs/analysis-pipeline.rst [hunk 4]
```diff
@@ -187,9 +187,10 @@ Generate figures
...
-folder (folder containing csv files) or specify this folder using ``-path-results`` flag. Figures will be created in the
-folder `results/`:
+folder (folder containing csv files) or specify this folder using ``-path-results`` flag. The flag ``-config`` is
+optional and an yml file containing the subjects/sites that want to be excluded have to be provided. Figures will 
+be created in the folder `results/`:
 
 .. code-block:: bash
 
-  sg_generate_figures -path-results ~/spineGeneric_results/results
+  sg_generate_figures -config PATH_TO_EXCLUDE.yml -path-results -path-results ~/spineGeneric_results/results
```
Comment (jcohenadad) [inline docs/analysis-pipeline.rst:196]:
```suggestion
  sg_generate_figures -exclude subjects_to_exclude.yml -path-results ~/spineGeneric_results/results
```

File: spinegeneric/cli/generate_figure.py [hunk 5]
```diff
@@ -215,11 +181,47 @@ def get_parameters():
         metavar='<dir_path>',
         help="Folder that includes all the output csv files (generated by process_data.sh). By default, takes the "
              "current folder.")
+    parser.add_argument(
+        '-config',
```
Comment (jcohenadad) [inline spinegeneric/cli/generate_figure.py:185]:
```suggestion
        '-exclude',
```

File: spinegeneric/cli/generate_figure.py [hunk 6]
```diff
@@ -215,11 +181,47 @@ def get_parameters():
...
+        required=False,
+        help=
+        "R|Config yaml file listing images that you want to remove from processing."
```
Comment (jcohenadad) [inline spinegeneric/cli/generate_figure.py:188]:
```suggestion
        "R|Yaml file listing subjects to remove from the statistics."
```

File: spinegeneric/cli/generate_figure.py [hunk 7]
```diff
@@ -215,11 +181,47 @@ def get_parameters():
...
+        "You can validate your yaml file at this website: http://www.yamllint.com/. Below is an example yaml file:\n"
```
Comment (jcohenadad) [inline spinegeneric/cli/generate_figure.py:189]:
```suggestion
        "Yaml file can be validated at this website: http://www.yamllint.com/. Below is an example yaml file:\n"
```

File: spinegeneric/cli/generate_figure.py [hunk 8]
```diff
@@ -246,7 +248,7 @@ def aggregate_per_site(dict_results, metric):
         # cluster values per site
         subject = fetch_subject(filename)
         # check if subject needs to be discarded
-        if not remove_subject(subject, metric):
+        if not remove_subject(subject, metric,dict_exclude_subj):
```
Comment (jcohenadad) [inline spinegeneric/cli/generate_figure.py:251]:
```suggestion
        if not remove_subject(subject, metric, dict_exclude_subj):
```

File: spinegeneric/cli/generate_figure.py [hunk 9]
```diff
@@ -469,6 +472,24 @@ def main():
     args = get_parameters()
     display_individual_subjects = args.indiv_subj
 
+    # create dict with subjects to exclude if input yml config file is passed
+    if args.config is not None:
```
Comment (jcohenadad) [inline spinegeneric/cli/generate_figure.py:476]:
```suggestion
    if args.exclude is not None:
```

File: spinegeneric/cli/generate_figure.py [hunk 10]
```diff
@@ -469,6 +472,24 @@ def main():
...
+        # check if input yml file exists
+        if os.path.isfile(args.config):
+            fname_yml = args.config
```
Comment (jcohenadad) [inline spinegeneric/cli/generate_figure.py:479]:
```suggestion
        if os.path.isfile(args.exclude):
            fname_yml = args.exclude
```

File: spinegeneric/cli/generate_figure.py [hunk 11]
```diff
@@ -501,16 +522,19 @@ def main():
...
         # Make it a pandas structure (easier for manipulations)
         df = pd.DataFrame.from_dict(results_dict, orient='index')
 
         # Compute statistics
-        if metric not in SITES_TO_EXCLUDE.keys():
+        if metric not in dict_exclude_subj.keys():
             sites_to_exclude = []
         else:
-            sites_to_exclude = SITES_TO_EXCLUDE[metric]
+            for subject in dict_exclude_subj[metric]:
```
Comment (jcohenadad) [inline spinegeneric/cli/generate_figure.py:534]: this functionality is not documented in the help, and using a variable called "*_subj" to refer to a site is confusing. I Would add a another variable (and flag) to remove a site.

**Review by jcohenadad — APPROVED** [review]

**Pull request merged**